File: Single_Number_III.cpp
class Solution {
public:
    /**
     * @param A : An integer array
     * @return : Two integers
     */
    vector<int> singleNumberIII(vector<int> &A) {
        // write your code here
        //get x_xor_y
        int x_xor_y = 0;
        for (auto& a : A){
            x_xor_y ^= a;
        }
        //get the lowest one in bit form
        int bits = x_xor_y & ~(x_xor_y - 1);
        int x = 0;
        for (auto& a : A){
            if (a & bits){
                x ^= a;
            }
        }
        return {x, x_xor_y ^ x};
        
    }
};

File: Single Number III.py
class Solution:
    """
    @param A : An integer array
    @return : Two integer
    """
    def singleNumberIII(self, A):
        # write your code here
        x_xor_y = 0
        for a in A:
            x_xor_y ^= a
        ## x_xor_y = x ^ y
        # diffBit = x_xor_y & (~x_xor_y + 1)
        diffBit = x_xor_y & (-1 * x_xor_y)
        x, y = 0, 0
        for a in A:
            if a & diffBit:
                x ^= a
            else:
                y ^= a
        return x, y
        